File: apps/desktop/src-tauri/src/downloads/heal.rs
//! Auto-réparation des à-côtés d'un téléchargement : snapshots (item.json,
//! affiches) et side-cars de sous-titres manquants pour les fichiers DÉJÀ
//! complets. Lancée en tâche de fond à chaque démarrage du moteur (en ligne) —
//! idempotente (tout ce qui existe est sauté), best-effort, et répare
//! notamment les téléchargements faits avant un correctif de récupération.

use super::{db, engine::Creds, fsops, meta, store, subs, trickplay};
use rusqlite::Connection;
use tauri::{AppHandle, Emitter};

fn fetch_json(agent: &ureq::Agent, url: &str, token: &str) -> Result<Vec<u8>, String> {
    use std::io::Read;
    let response = agent
        .get(url)
        .set("X-Emby-Token", token)
        .call()
        .map_err(|e| format!("GET: {e}"))?;
    let mut bytes = Vec::new();
    response
        .into_reader()
        .take(8 * 1024 * 1024)
        .read_to_end(&mut bytes)
        .map_err(|e| format!("read: {e}"))?;
    Ok(bytes)
}

fn complete_files(conn: &Connection) -> Vec<(String, String, Option<String>)> {
    let Ok(mut stmt) = conn.prepare(
        "SELECT DISTINCT item_id, media_source_id, subtitles_json
         FROM files WHERE status = 'complete'",
    ) else {
        return Vec::new();
    };
    let Ok(rows) = stmt.query_map([], |row| {
        Ok((row.get(0)?, row.get(1)?, row.get(2)?))
    }) else {
        return Vec::new();
    };
    rows.flatten().collect()
}

/// Corps de la réparation (thread de fond). Retourne le nombre d'items touchés.
pub fn run(app: &AppHandle, creds: &Creds) -> usize {
    let Ok(db_path) = db::db_path(app) else { return 0 };
    let Ok(conn) = db::open(&db_path) else { return 0 };
    let Ok(root) = fsops::resolve_root(app) else { return 0 };
    let agent = ureq::AgentBuilder::new()
        .timeout(std::time::Duration::from_secs(20))
        .build();

    let mut healed = 0;
    for (item_id, media_source_id, subtitles_json) in complete_files(&conn) {
        let mut touched = false;
        // Snapshot absent, affiche de série manquante, ou snapshot d'une
        // version antérieure (sans segments.json / DTO enrichi) : un
        // re-snapshot complet répare tout, il saute ce qui est déjà en place.
        if let Ok(Some(spec)) = meta::get_spec(&conn, &item_id) {
            let missing_series_poster =
                spec.series_id.is_some() && !meta::series_primary_exists(&root, &item_id);
            let outdated_meta = meta::meta_version(&conn, &item_id) < meta::CURRENT_META_VERSION;
            let needs_snapshot = !meta::snapshot_exists(&root, &item_id)
                || missing_series_poster
                || outdated_meta;
            if needs_snapshot
                && meta::snapshot(&agent, &creds.server_url, &creds.token, &root, &conn, &spec)
                    .is_ok()
            {
                touched = true;
            }
        }
        // Trickplay manquant (téléchargements d'avant ce correctif) : récupérer
        // le manifeste (fields=Trickplay) puis les planches.
        if !trickplay::exists(&root, &item_id) {
            let url = format!("{}/api/jellyfin/Items/{item_id}?fields=Trickplay", creds.server_url);
            if let Ok(item_json) = fetch_json(&agent, &url, &creds.token) {
                let msrc = store::first_media_source_id(&conn, &item_id)
                    .unwrap_or_else(|| item_id.clone());
                let saved = trickplay::download(
                    &agent, &creds.server_url, &creds.token, &root, &item_id, &msrc, &item_json,
                );
                touched = touched || saved > 0;
            }
        }
        if let Some(json) = &subtitles_json {
            let specs = subs::parse_specs(json);
            if !specs.is_empty() {
                // fetch_all saute les fichiers déjà présents (idempotent).
                let fetched = subs::fetch_all(
                    &agent, &creds.server_url, &creds.token, &root, &item_id,
                    &media_source_id, &specs,
                );
                touched = touched || fetched > 0;
            }
        }
        if touched {
            healed += 1;
        }
    }
    if healed > 0 {
        let _ = app.emit(super::engine::EVENT_CHANGED, ());
    }
    healed
}

/// Variante détachée (démarrage du moteur).
pub fn spawn(app: AppHandle, creds: Creds) {
    std::thread::spawn(move || {
        let _ = run(&app, &creds);
    });
}

File: apps/desktop-electron/src/main/downloads/heal.ts
/**
 * Auto-réparation des à-côtés d'un téléchargement.
 *
 * Snapshots, affiches et side-cars manquants pour les fichiers DÉJÀ complets.
 * Lancée en tâche de fond à chaque démarrage du moteur, en ligne. Idempotente —
 * tout ce qui existe est sauté —, best-effort, et c'est elle qui rattrape les
 * téléchargements faits avant un correctif de récupération.
 *
 * Portage de `apps/desktop/src-tauri/src/downloads/heal.rs`.
 */

import type { DatabaseSync } from "node:sqlite";
import type { FetchBytes } from "./fetcher";
import { MAX_JSON_BYTES } from "./fetcher";
import { parseJson } from "./json";
import {
  CURRENT_META_VERSION,
  getSpec,
  metaVersion,
  seriesPrimaryExists,
  snapshotExists,
} from "./meta";
import { snapshot } from "./snapshot";
import { firstMediaSourceId } from "./store";
import { fetchAll, parseSpecs } from "./subs";
import { text, textOrNull } from "./rows";
import * as trickplay from "./trickplay";

interface Complet {
  itemId: string;
  mediaSourceId: string;
  subtitlesJson: string | null;
}

function complets(db: DatabaseSync): Complet[] {
  return db
    .prepare(
      `SELECT DISTINCT item_id, media_source_id, subtitles_json
       FROM files WHERE status = 'complete'`,
    )
    .all()
    .map((row) => ({
      itemId: text(row, "item_id"),
      mediaSourceId: text(row, "media_source_id"),
      subtitlesJson: textOrNull(row, "subtitles_json"),
    }));
}

/**
 * Répare ce qui manque. Retourne le nombre d'items touchés.
 *
 * Ne lève jamais : elle tourne en fond, et un item récalcitrant ne doit pas
 * empêcher les suivants d'être réparés.
 */
export async function heal(
  fetchBytes: FetchBytes,
  db: DatabaseSync,
  serverUrl: string,
  root: string,
  nowMs: number,
): Promise<number> {
  let repares = 0;

  for (const item of complets(db)) {
    let touche = false;

    // Snapshot absent, affiche de série manquante, ou snapshot d'une version
    // antérieure (sans segments ni DTO enrichi) : un re-snapshot complet répare
    // tout, et il saute ce qui est déjà en place.
    const spec = getSpec(db, item.itemId);
    if (spec !== null) {
      const afficheSerieManquante =
        spec.seriesId !== null && !seriesPrimaryExists(root, item.itemId);
      const versionDepassee = metaVersion(db, item.itemId) < CURRENT_META_VERSION;
      if (!snapshotExists(root, item.itemId) || afficheSerieManquante || versionDepassee) {
        try {
          await snapshot(fetchBytes, db, serverUrl, root, spec, nowMs);
          touche = true;
        } catch {
          // Item non réparé ce tour-ci ; on continue avec les suivants.
        }
      }
    }

    // Trickplay manquant : récupérer le manifeste puis les planches.
    //
    // Le marqueur `trickplay.none` est ce qui empêche cette requête de
    // repartir à CHAQUE démarrage pour chaque item que le serveur ne sait pas
    // illustrer. Sans lui, un catalogue de cinquante films sans planches
    // faisait cinquante appels au lancement, en série, à chaque fois.
    if (
      !trickplay.exists(root, item.itemId) &&
      !trickplay.noneRecently(root, item.itemId, nowMs)
    ) {
      const itemJson = await fetchBytes(
        `${serverUrl}/api/jellyfin/Items/${item.itemId}?fields=Trickplay`,
        MAX_JSON_BYTES,
      );
      if (itemJson !== null) {
        const msrc = firstMediaSourceId(db, item.itemId) ?? item.itemId;
        const planches = await trickplay.download(
          fetchBytes,
          serverUrl,
          root,
          item.itemId,
          msrc,
          parseJson(itemJson),
          nowMs,
        );
        touche = touche || planches > 0;
      }
    }

    if (item.subtitlesJson !== null) {
      const specs = parseSpecs(item.subtitlesJson);
      if (specs.length > 0) {
        // `fetchAll` saute les fichiers déjà présents : rappeler est gratuit.
        const obtenus = await fetchAll(
          fetchBytes,
          serverUrl,
          root,
          item.itemId,
          item.mediaSourceId,
          specs,
        );
        touche = touche || obtenus > 0;
      }
    }

    if (touche) repares += 1;
  }

  return repares;
}
